File: kakao.app/js/ked/view/colmns/funtree.js
var _k_ = {empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, extend: function (c,p) {for (var k in p) { if (Object.hasOwn(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, clone: function (o,v) { v ??= new Map(); if (Array.isArray(o)) { if (!v.has(o)) {var r = []; v.set(o,r); for (var i=0; i < o.length; i++) {if (!v.has(o[i])) { v.set(o[i],_k_.clone(o[i],v)) }; r.push(v.get(o[i]))}}; return v.get(o) } else if (typeof o == 'string') { if (!v.has(o)) {v.set(o,''+o)}; return v.get(o) } else if (o != null && typeof o == 'object' && o.constructor.name == 'Object') { if (!v.has(o)) { var k, r = {}; v.set(o,r); for (k in o) { if (!v.has(o[k])) { v.set(o[k],_k_.clone(o[k],v)) }; r[k] = v.get(o[k]) }; }; return v.get(o) } else {return o} }}

var funSyntax, funtree, SYMBOL

import kxk from "../../../kxk.js"
let post = kxk.post
let slash = kxk.slash

import nfs from "../../../kxk/nfs.js"

import color from "../../theme/color.js"
import theme from "../../theme/theme.js"
import icons from "../../theme/icons.js"

import belt from "../../edit/tool/belt.js"

import choices from "../menu/choices.js"

SYMBOL = {clss:'■',unbound:'->',bound:'=>',async_unbound:'○→',async_bound:'●→'}

funSyntax = (function ()
{
    function funSyntax (tree)
    {
        this.tree = tree
    
        this["getChar"] = this["getChar"].bind(this)
        this["getColor"] = this["getColor"].bind(this)
        this["getClass"] = this["getClass"].bind(this)
        this["setSegls"] = this["setSegls"].bind(this)
        this["setLines"] = this["setLines"].bind(this)
        this["clear"] = this["clear"].bind(this)
        this.color = {class:theme.funtree.class,async:theme.funtree.async,bound:theme.funtree.bound,bound_async:theme.funtree.bound_async,func:theme.funtree.func,func_async:theme.funtree.func_async,test0:theme.funtree.test0,test1:theme.funtree.test1,test2:theme.funtree.test2,test3:theme.funtree.test3}
    }

    funSyntax.prototype["clear"] = function ()
    {}

    funSyntax.prototype["setLines"] = function (lines)
    {
        console.log('setLines')
    }

    funSyntax.prototype["setSegls"] = function (segls)
    {}

    funSyntax.prototype["getClass"] = function (x, y)
    {
        return ''
    }

    funSyntax.prototype["getColor"] = function (x, y)
    {
        var char, clr, item, name

        item = this.tree.items[y]
        name = item.name
        char = name[x]
        if (char === ' ')
        {
            return [0,0,0]
        }
        if (item.clss)
        {
            clr = this.color.class
        }
        else if (item.test)
        {
            clr = this.color['test' + parseInt(belt.numIndent(item.name) / 2)]
        }
        else if (item.async)
        {
            clr = (item.bound ? this.color.bound_async : this.color.func_async)
        }
        else if (item.bound)
        {
            clr = this.color.bound
        }
        else
        {
            clr = this.color.func
        }
        if (item.static)
        {
            clr = color.brighten(clr,0.2)
        }
        else if (char === '@' && _k_.empty(name[x + 1]))
        {
            clr = this.color.class
        }
        if (char === '▸')
        {
            clr = color.darken(clr,0.5)
        }
        else if (char === '@')
        {
            clr = color.darken(clr,((clr === this.color.class ? 0.75 : 0.5)))
        }
        else if (char === SYMBOL.clss)
        {
            clr = color.darken(clr,0.2)
        }
        else if (_k_.in(char,SYMBOL.bound) && item.bound)
        {
            clr = color.darken(clr,0.5)
        }
        else if (_k_.in(char,SYMBOL.unbound))
        {
            clr = color.darken(clr,0.7)
        }
        return clr
    }

    funSyntax.prototype["getChar"] = function (x, y, char)
    {
        return char
    }

    return funSyntax
})()


funtree = (function ()
{
    _k_.extend(funtree, choices)
    function funtree (editor, name, features)
    {
        this.editor = editor
    
        this["onFileIndexed"] = this["onFileIndexed"].bind(this)
        this["lineIndexForFunc"] = this["lineIndexForFunc"].bind(this)
        this["lineIndexOfPrevFunc"] = this["lineIndexOfPrevFunc"].bind(this)
        this["lineIndexOfNextFunc"] = this["lineIndexOfNextFunc"].bind(this)
        this["lineIndexOfCurrentFunc"] = this["lineIndexOfCurrentFunc"].bind(this)
        this["nameOfCurrentFunc"] = this["nameOfCurrentFunc"].bind(this)
        this["selectItemForLineIndex"] = this["selectItemForLineIndex"].bind(this)
        this["onCursorsSet"] = this["onCursorsSet"].bind(this)
        funtree.__super__.constructor.call(this,this.editor.screen,name,features)
        this.state.syntax = new funSyntax(this)
        this.editor.state.on('cursorsSet',this.onCursorsSet)
        post.on('file.loaded',this.clear)
        post.on('file.indexed',this.onFileIndexed)
    }

    funtree.prototype["onCursorsSet"] = function ()
    {
        return this.selectItemForLineIndex(this.editor.state.mainCursor()[1])
    }

    funtree.prototype["selectItemForLineIndex"] = function (li)
    {
        var idx, item

        var list = _k_.list(this.items)
        for (idx = 0; idx < list.length; idx++)
        {
            item = list[idx]
            if (item.line - 1 <= li && (idx >= this.items.length - 1 || this.items[idx + 1].line - 1 > li))
            {
                this.state.setSelections([belt.rangeOfLine(this.state.s.lines,idx)])
                this.state.setMainCursor(0,idx)
                return
            }
        }
    }

    funtree.prototype["nameOfCurrentFunc"] = function ()
    {
        return this.current().name
    }

    funtree.prototype["lineIndexOfCurrentFunc"] = function ()
    {
        return this.current().line - 1
    }

    funtree.prototype["lineIndexOfNextFunc"] = function ()
    {
        return (this.items[this.currentIndex() + 1] != null ? this.items[this.currentIndex() + 1].line : undefined) - 1
    }

    funtree.prototype["lineIndexOfPrevFunc"] = function ()
    {
        return (this.items[this.currentIndex() - 1] != null ? this.items[this.currentIndex() - 1].line : undefined) - 1
    }

    funtree.prototype["lineIndexForFunc"] = function (func)
    {
        var item

        var list = _k_.list(this.items)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            item = list[_a_]
            if (item.name === func.name && item.class === func.class)
            {
                return item.line - 1
            }
        }
        if (func.name === `   -> ${func.class}` && func.line)
        {
            return this.lineIndexForFunc({class:func.class,name:'   -> @'})
        }
        if (func.name === "   -> @" && func.line)
        {
            return this.lineIndexForFunc({class:func.class,name:`   -> ${func.class}`})
        }
        console.log("can't find func:",func)
    }

    funtree.prototype["onFileIndexed"] = function (path, info)
    {
        var clss, clssl, func, funcs, indt, items, name, symbol, _158_22_

        if (path !== this.editor.currentFile)
        {
            return
        }
        if (_k_.empty(info.classes) && _k_.empty(info.funcs))
        {
            return
        }
        clssl = _k_.clone(info.classes)
        funcs = _k_.clone(info.funcs)
        var list = _k_.list(clssl)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            clss = list[_a_]
            clss.file = ((_158_22_=clss.file) != null ? _158_22_ : path)
            clss.name = ' ' + SYMBOL.clss + ' ' + clss.name
        }
        var list1 = _k_.list(funcs)
        for (var _b_ = 0; _b_ < list1.length; _b_++)
        {
            func = list1[_b_]
            if (func.test)
            {
                func.name = ' ' + belt.reindent(4,2,func.name)
            }
            else
            {
                if (func.async)
                {
                    if (func.bound)
                    {
                        symbol = SYMBOL.async_bound
                    }
                    else
                    {
                        symbol = SYMBOL.async_unbound
                    }
                }
                else
                {
                    if (func.bound)
                    {
                        symbol = SYMBOL.bound
                    }
                    else
                    {
                        symbol = SYMBOL.unbound
                    }
                }
                name = func.name
                if (func.static)
                {
                    name = '@' + name
                }
                indt = (func.class ? '   ' : ' ')
                func.name = indt + symbol + ' ' + name
            }
        }
        items = clssl.concat(funcs)
        items.sort(function (a, b)
        {
            return a.line - b.line
        })
        this.set(items,'name')
        post.emit('funtree.loaded')
        this.onCursorsSet()
        return post.emit('redraw')
    }

    funtree.prototype["emitAction"] = function (action, choice, event)
    {
        if (_k_.empty(choice))
        {
            console.error('funtree.emitAction -- empty choice ▸ action:',action)
            console.error('funtree.emitAction -- empty choice ▸ event:',event)
            return
        }
        switch (action)
        {
            case 'right':
                post.emit('goto.line',choice.line - 1,'ind','topDelta')
                return

            case 'click':
            case 'return':
                post.emit('goto.line',choice.line - 1,'ind','topDelta')
                post.emit('focus','editor')
                return

            case 'drag':
                post.emit('goto.line',choice.line - 1,'ind','topDelta')
                return

        }

        return funtree.__super__.emitAction.call(this,action,choice,event)
    }

    return funtree
})()

export default funtree;

File: kakao.app/love/lua/view/colmns/funtree.lua
--    ████████  ███   ███  ███   ███  █████████  ████████   ████████  ████████  
--    ███       ███   ███  ████  ███     ███     ███   ███  ███       ███       
--    ██████    ███   ███  ███ █ ███     ███     ███████    ███████   ███████   
--    ███       ███   ███  ███  ████     ███     ███   ███  ███       ███       
--    ███        ███████   ███   ███     ███     ███   ███  ████████  ████████  

-- use ../../../kxk    ▪ post slash
-- use ../../../kxk    ◆ nfs
-- use ../../theme     ◆ color theme icons
-- use ../../edit/tool ◆ belt
-- use ../menu         ◆ choices 

--[[
     ███████  ███   ███  ███   ███  █████████   ███████   ███   ███
    ███        ███ ███   ████  ███     ███     ███   ███   ███ ███ 
    ███████     █████    ███ █ ███     ███     █████████    █████  
         ███     ███     ███  ████     ███     ███   ███   ███ ███ 
    ███████      ███     ███   ███     ███     ███   ███  ███   ███
--]]

local SYMBOL = {
    clss = '■', -- '■' , 
    unbound = '->', -- '▸', 
    bound = '->', -- '▸', 
    async_unbound = '○→', -- '○', 
    async_bound = '●→', -- '●' 
    }


local funSyntax = class("funSyntax")
    


function funSyntax:init(tree) 
        self.tree = tree
        
        self.color = {
            class = theme.funtree.class, 
            async = theme.funtree.async, 
            bound = theme.funtree.bound, 
            bound_async = theme.funtree.bound_async, 
            func = theme.funtree.func, 
            func_async = theme.funtree.func_async, 
            test0 = theme.funtree.test0, 
            test1 = theme.funtree.test1, 
            test2 = theme.funtree.test2, 
            test3 = theme.funtree.test3
            }
        return self
    end


function funSyntax:clear() 
    
    end


function funSyntax:setLines(lines) 
    return print('setLines')
    end

function funSyntax:setSegls(segls) 
    
    end

--  ███████   ███████   ███       ███████   ████████ 
-- ███       ███   ███  ███      ███   ███  ███   ███
-- ███       ███   ███  ███      ███   ███  ███████  
-- ███       ███   ███  ███      ███   ███  ███   ███
--  ███████   ███████   ███████   ███████   ███   ███


function funSyntax:getClass(x, y) 
    return ''
    end

function funSyntax:getColor(x, y) 
        local item = self.tree.items[y]
        local name = item.name
        local char = name[x]
        if (char == ' ') then return array(0, 0, 0) end
        
        if item.clss then clr = self.color.class
        elseif item.test then clr = self.color[('test' + int((belt.numIndent(item.name) / 2)))]
        elseif item.async then clr = ((item.bound and self.color.bound_async) or self.color.func_async)
        elseif item.bound then clr = self.color.bound
        else clr = self.color.func
        end
        
        if item.static then clr = color.brighten(clr, 0.2)
        elseif ((char == '@') and empty(name[(x + 1)])) then 
                              local clr = self.color.class
        end
        
        if (char == '▸') then clr = color.darken(clr, 0.5)
        elseif (char == '@') then clr = color.darken(clr((((clr == self.color.class) and 0.75) or 0.5)))
        elseif (char == SYMBOL.clss) then clr = color.darken(clr, 0.2)
        end
        
        return clr
    end


function funSyntax:getChar(x, y, char) 
    return char
    end

--[[    
    ████████  ███   ███  ███   ███  █████████  ████████   ████████  ████████
    ███       ███   ███  ████  ███     ███     ███   ███  ███       ███     
    ██████    ███   ███  ███ █ ███     ███     ███████    ███████   ███████ 
    ███       ███   ███  ███  ████     ███     ███   ███  ███       ███     
    ███        ███████   ███   ███     ███     ███   ███  ████████  ████████
--]]


local funtree = class("funtree", choices)
    


function funtree:init(editor, name, features) 
        self.editor = editor
        
        choices.init(self, name, features)
        
        self.state.syntax = funSyntax(self)
        
        self.editor.state:on('cursorsSet', self.onCursorsSet, self)
        
        post:on('file.loaded', self.clear, self)
        post:on('file.indexed', self.onFileIndexed, self)
        return self
    end


function funtree:onCursorsSet() 
        return self:selectItemForLineIndex(self.editor.state:mainCursor()[1])
    end


function funtree:selectItemForLineIndex(li) 
        for idx, item in ipairs(self.items) do 
            if (((item.line - 1) <= li) and ((idx >= (#self.items - 1)) or ((self.items[(idx + 1)].line - 1) > li))) then 
                self.state:setSelections(array(belt.rangeOfLine(self.state.s.lines, idx)))
                self.state:setMainCursor(1, idx)
                return
            end
        end
    end


function funtree:nameOfCurrentFunc() 
    return self:current().name
    end

function funtree:lineIndexOfCurrentFunc() 
    return (self:current().line - 1)
    end

function funtree:lineIndexOfNextFunc() 
    return (self.items[(self:currentIndex() + 1)].line - 1)
    end

function funtree:lineIndexOfPrevFunc() 
    return (self.items[(self:currentIndex() - 1)].line - 1)
    end

-- ███      ███  ███   ███  ████████        ████████   ███████   ████████         ████████  ███   ███  ███   ███   ███████
-- ███      ███  ████  ███  ███             ███       ███   ███  ███   ███        ███       ███   ███  ████  ███  ███     
-- ███      ███  ███ █ ███  ███████         ██████    ███   ███  ███████          ██████    ███   ███  ███ █ ███  ███     
-- ███      ███  ███  ████  ███             ███       ███   ███  ███   ███        ███       ███   ███  ███  ████  ███     
-- ███████  ███  ███   ███  ████████        ███        ███████   ███   ███        ███        ███████   ███   ███   ███████


function funtree:lineIndexForFunc(func) 
        for _, item in ipairs(self.items) do 
            if ((item.name == func.name) and (item.class == func.class)) then 
                return (item.line - 1)
            end
        end
        
        local funcname = "   -> " .. tostring(func.class) .. ""
        if ((func.name == funcname) and func.line) then 
            return self:lineIndexForFunc({class = func.class, name = '   -> @'})
        end
        
        if ((func.name == "   -> @") and func.line) then 
            return self:lineIndexForFunc({class = func.class, name = funcname})
        end
        
        -- log "can't find func: items:" @items
        return print("can't find func:", function () end)
    end

--  ███████   ███   ███        ███  ███   ███  ███████    ████████  ███   ███
-- ███   ███  ████  ███        ███  ████  ███  ███   ███  ███        ███ ███ 
-- ███   ███  ███ █ ███        ███  ███ █ ███  ███   ███  ███████     █████  
-- ███   ███  ███  ████        ███  ███  ████  ███   ███  ███        ███ ███ 
--  ███████   ███   ███        ███  ███   ███  ███████    ████████  ███   ███


function funtree:onFileIndexed(path, info) 
        if (path ~= self.editor.currentFile) then return end
        if empty((info.classes and empty), info.funcs) then return end
        
        local clssl = clone(info.classes)
        local funcs = clone(info.funcs)
        
        for _, clss in ipairs(clssl) do 
            clss.file = clss.file or path
            clss.name = ' ' .. SYMBOL.clss .. (' ' + clss.name)
        end
        
        for _, func in ipairs(funcs) do 
            if func.test then 
                func.name = ' ' .. belt.reindent(4, 2, func.name)
            else 
                if func.async then 
                    if func.bound then 
                        local symbol = SYMBOL.async_bound
                    else 
                        local symbol = SYMBOL.async_unbound
                    end
                else 
                    if func.bound then 
                        local symbol = SYMBOL.bound
                    else 
                        local symbol = SYMBOL.unbound
                    end
                end
                
                local name = func.name
                if func.static then 
                    name = ('@' + name)
                end
                
                local indt = (function () 
    if func.class then 
    return '   ' else 
    return ' '
                       end
end)()
                func.name = indt .. symbol .. ' ' .. name
            end
        end
        
        local items = (clssl + funcs)
        items:sort(function (a, b) 
    return (a.line - b.line)
end)
        
        self:set(items, 'name')
        post:emit('funtree.loaded')
        self:onCursorsSet()
        return post:emit('redraw')
    end

--  0000000    0000000  000000000  000   0000000   000   000  
-- 000   000  000          000     000  000   000  0000  000  
-- 000000000  000          000     000  000   000  000 0 000  
-- 000   000  000          000     000  000   000  000  0000  
-- 000   000   0000000     000     000   0000000   000   000  


function funtree:emitAction(action, choice, event) 
        if empty(choice) then 
            error('funtree.emitAction -- empty choice ▸ action:', action)
            error('funtree.emitAction -- empty choice ▸ event:', event)
            return
        end
        
        if (action == 'right') then 
                post:emit('goto.line', (choice.line - 1), 'ind', 'topDelta')
                return
        elseif (action == 'click') or (action == 'return') then 
                post:emit('goto.line', (choice.line - 1), 'ind', 'topDelta')
                post:emit('focus', 'editor')
                return
        elseif (action == 'drag') then 
                post:emit('goto.line', (choice.line - 1), 'ind', 'topDelta')
                return
        end
        
        return choices.emitAction(self, action, choice, event)
    end

return funtree